### PR TITLE
Add UsersAdditions

### PR DESCRIPTION
**Pull request by actimellka**

### CODE REVIEW

File: src/UserService.Business/Commands/User/EditUserCommand.cs [hunk 1]
```diff
@@ -1,105 +1,140 @@
...
+      Guid requestSenderId = _httpContextAccessor.HttpContext.GetUserId();
+
+      bool isAdmin = await _accessValidator.IsAdminAsync(userId);
+      bool isAddEditRemoveUsers = await _accessValidator.HasRightsAsync(Rights.AddEditRemoveUsers);
+
+      if (!isAddEditRemoveUsers && userId != requestSenderId)
+      {
+        _responseCreator.CreateFailureResponse<bool>(HttpStatusCode.Forbidden);
+      }
+
```
Comment (MariyaSin) [inline src/UserService.Business/Commands/User/EditUserCommand.cs:69]: you make the ckeck in  79 line. remove it from here

File: src/UserService.Business/Commands/User/EditUserCommand.cs [hunk 2]
```diff
@@ -1,105 +1,140 @@
...
+        o => o.path.EndsWith(nameof(EditUserRequest.IsActive), StringComparison.OrdinalIgnoreCase));
+
+      Operation<EditUserRequest> isAdminOperation = patch.Operations.FirstOrDefault(
+        o => o.path.EndsWith(nameof(EditUserRequest.IsAdmin), StringComparison.OrdinalIgnoreCase));
+
+      Operation<EditUserRequest> isGenderOperation = patch.Operations.FirstOrDefault(
+        o => o.path.EndsWith(nameof(EditUserRequest.GenderId), StringComparison.OrdinalIgnoreCase));
+
+      if ((!isAddEditRemoveUsers && userId != requestSenderId) ||
+        (isActiveOperation is not null && !isAddEditRemoveUsers) ||
```
Comment (MariyaSin) [inline src/UserService.Business/Commands/User/EditUserCommand.cs:80]: is active can change AddEditRemoveUsers and admin

File: src/UserService.Business/Commands/User/EditUserCommand.cs [hunk 3]
```diff
@@ -1,105 +1,140 @@
...
+        response.Body = await _userRepository.EditUserAdditionAsync(userId, dbUserAdditionPatch);
+
+        if (!response.Body)
+        {
+          return _responseCreator.CreateFailureResponse<bool>(HttpStatusCode.NotFound);
+        }
+      }
+
+      await _cacheNotebook.RemoveAsync(userId);
+
```
Comment (MariyaSin) [inline src/UserService.Business/Commands/User/EditUserCommand.cs:130]: anyway you get empty errors list.  remove 96-98 lines
something strange in 100 line, remove it

File: src/UserService.Data.Provider.MsSql.Ef/Migrations/20211213225100_CreateUsersAdditionsTableAndSomeUsersColumnsDrop.cs [hunk 4]
```diff
@@ -0,0 +1,55 @@
+﻿using LT.DigitalOffice.UserService.Models.Db;
+using Microsoft.EntityFrameworkCore.Infrastructure;
+using Microsoft.EntityFrameworkCore.Migrations;
+using System;
+
+namespace LT.DigitalOffice.UserService.Data.Provider.MsSql.Ef.Migrations
+{
+  [DbContext(typeof(UserServiceDbContext))]
+  [Migration("20211213225100_CreateUsersAdditionsTable")]
```
Comment (MariyaSin) [inline src/UserService.Data.Provider.MsSql.Ef/Migrations/20211213225100_CreateUsersAdditionsTableAndSomeUsersColumnsDrop.cs:9]: the sting must contain cs file name.

File: src/UserService.Mappers/Db/DbUserMapper.cs [hunk 5]
```diff
@@ -1,97 +1,122 @@
...
+      Guid userId = Guid.NewGuid();
+      Guid createdBy = _httpContextAccessor.HttpContext.GetUserId();
+      DateTime createdAtUtc = DateTime.UtcNow;
+
+      DbUser dbUser = new()
+      {
+        Id = userId,
+        FirstName = request.FirstName.Trim(),
+        LastName = request.LastName.Trim(),
+        MiddleName = !string.IsNullOrEmpty(request.MiddleName?.Trim()) ? request.MiddleName.Trim() : null,
```
Comment (MariyaSin) [inline src/UserService.Mappers/Db/DbUserMapper.cs:39]: no need trim in condition

Comment (MariyaSin) [reply src/UserService.Mappers/Db/DbUserMapper.cs:39]: same in 88 line

File: src/UserService.Mappers/Models/UserInfoMapper.cs [hunk 6]
```diff
@@ -1,50 +1,47 @@
...
+        return null;
+      }
+
+      return new UserInfo
+      {
+        Id = dbUser.Id,
+        FirstName = dbUser.FirstName,
+        LastName = dbUser.LastName,
+        MiddleName = dbUser.MiddleName,
+
```
Comment (MariyaSin) [inline src/UserService.Mappers/Models/UserInfoMapper.cs:32]: do you need empty line here?

File: src/UserService.Models.Db/DbUserAddition.cs [hunk 7]
```diff
@@ -1,40 +1,36 @@
...
+    public double? Latitude { get; set; }
+    public double? Longitude { get; set; }
+    public DateTime? BusinessHoursFromUtc { get; set; }
+    public DateTime? BusinessHoursToUtc { get; set; }
+    public Guid ModifiedBy { get; set; }
+    public DateTime ModifiedAtUtc { get; set; }
+    public DbUser User { get; set; }
+  }
+
+  public class UserAdditionConfiguration : IEntityTypeConfiguration<DbUserAddition>
```
Comment (MariyaSin) [inline src/UserService.Models.Db/DbUserAddition.cs:25]: DbUser....

File: src/UserService.Models.Db/DbUserAddition.cs [hunk 8]
```diff
@@ -1,40 +1,36 @@
...
+  {
+    public void Configure(EntityTypeBuilder<DbUserAddition> builder)
+    {
+      builder
+        .HasKey(x => x.Id);
+
+      builder
```
Comment (MariyaSin) [inline src/UserService.Models.Db/DbUserAddition.cs:32]: remove it before .HasKey

File: src/UserService.Models.Db/DbUserAddition.cs [hunk 9]
```diff
@@ -1,40 +1,36 @@
...
+        .ToTable(DbUserAddition.TableName);
```
Comment (MariyaSin) [inline src/UserService.Models.Db/DbUserAddition.cs:33]: Has one Use with one UserAddition

File: src/UserService.Validation/User/EditUserRequestValidator.cs [hunk 10]
```diff
@@ -1,169 +1,252 @@
...
+      #endregion
+
+      #region GenderId 
+
+      AddFailureForPropertyIf(
+        nameof(EditUserRequest.GenderId),
+        x => x == OperationType.Replace,
+        new Dictionary<Func<Operation<EditUserRequest>, bool>, string>
+        {
+          { x => Guid.TryParse(x.value?.ToString(), out Guid _) || String.IsNullOrEmpty(x.value.ToString()), "Incorrect format of GenderId." }, //проверить на null
```
Comment (MariyaSin) [inline src/UserService.Validation/User/EditUserRequestValidator.cs:215]: String.IsNullOrEmpty(x.value?.ToString())
remove comment in end of line

File: src/UserService.Validation/User/EditUserRequestValidator.cs [hunk 11]
```diff
@@ -1,169 +1,252 @@
...
+
+      #region BusinessHoursFromUtc
+
+      AddFailureForPropertyIf(
+      nameof(EditUserRequest.BusinessHoursFromUtc), 
+      x => x == OperationType.Replace, 
+      new Dictionary<Func<Operation<EditUserRequest>, bool>, string>
+      {
+        { x => string.IsNullOrEmpty(x.value?.ToString())? true :
+          DateTime.TryParse(x.value.ToString(), out DateTime result),
```
Comment (MariyaSin) [inline src/UserService.Validation/User/EditUserRequestValidator.cs:154]: check what value has time format

File: src/UserService.Validation/User/EditUserRequestValidator.cs [hunk 12]
```diff
@@ -1,169 +1,252 @@
...
+        "Incorrect format of BusinessHoursFromUtc."
+        },
+        {
+          x => DateTime.TryParse("hh:mm tt", out DateTime result),
+          "Incorrect format of BusinessHoursFromUtc."
+        }
+      });
```
Comment (MariyaSin) [inline src/UserService.Validation/User/EditUserRequestValidator.cs:161]: add cascade. If request value is null you get false in second check

Comment (MariyaSin) [reply src/UserService.Validation/User/EditUserRequestValidator.cs:161]: split checks

File: src/UserService.Validation/User/EditUserRequestValidator.cs [hunk 13]
```diff
@@ -1,169 +1,252 @@
...
+      x => x == OperationType.Replace,
+      new Dictionary<Func<Operation<EditUserRequest>, bool>, string>
+      {
+        { x => string.IsNullOrEmpty(x.value?.ToString())? true :
+          DateTime.TryParse(x.value.ToString(), out DateTime result),
+        "Incorrect format of BusinessHoursToUtc."
+        },
+        {
+          x => DateTime.TryParse("hh:mm tt", out DateTime result),
+          "Incorrect format of BusinessHoursToUtc."
```
Comment (MariyaSin) [inline src/UserService.Validation/User/EditUserRequestValidator.cs:178]: same

File: src/UserService.Business/Commands/User/EditUserCommand.cs [hunk 14]
```diff
@@ -1,105 +1,124 @@
...
+      Operation<EditUserRequest> isActiveOperation = patch.Operations.FirstOrDefault(
+        o => o.path.EndsWith(nameof(EditUserRequest.IsActive), StringComparison.OrdinalIgnoreCase));
+
+      Operation<EditUserRequest> isAdminOperation = patch.Operations.FirstOrDefault(
+        o => o.path.EndsWith(nameof(EditUserRequest.IsAdmin), StringComparison.OrdinalIgnoreCase));
+
+      Operation<EditUserRequest> isGenderOperation = patch.Operations.FirstOrDefault(
+        o => o.path.EndsWith(nameof(EditUserRequest.GenderId), StringComparison.OrdinalIgnoreCase));
+
+      if ((!isAddEditRemoveUsers && userId != requestSenderId) ||
```
Comment (MariyaSin) [inline src/UserService.Business/Commands/User/EditUserCommand.cs:74]: && !isAdmin

File: src/UserService.Business/Commands/User/EditUserCommand.cs [hunk 15]
```diff
@@ -1,105 +1,124 @@
...
+        (isActiveOperation is not null && (!isAddEditRemoveUsers || !isAdmin)) ||
```
Comment (MariyaSin) [inline src/UserService.Business/Commands/User/EditUserCommand.cs:75]: (isActiveOperation is not null && !(isAddEditRemoveUsers || isAdmin))

File: src/UserService.Business/Commands/User/EditUserCommand.cs [hunk 16]
```diff
@@ -1,105 +1,124 @@
...
-
-        bool switchActiveStatusResult = await _credentialsRepository.SwitchActiveStatusAsync(userId, newValue);
-
-        if (!switchActiveStatusResult)
-        {
-          response.Errors.Add("Can not change is active status.");
-        }
-        else if (!newValue)
-        {
-          await _bus.Publish<IDisactivateUserRequest>(IDisactivateUserRequest.CreateObj(
```
Comment (MariyaSin) [inline src/UserService.Business/Commands/User/EditUserCommand.cs:86]: bring it back when 'if (switchActiveStatusResult)..' to 88 line

File: src/UserService.Validation/User/EditUserRequestValidator.cs [hunk 17]
```diff
@@ -1,169 +1,244 @@
...
+      #endregion
+
+      #region GenderId 
+
+      AddFailureForPropertyIf(
+        nameof(EditUserRequest.GenderId),
+        x => x == OperationType.Replace,
+        new Dictionary<Func<Operation<EditUserRequest>, bool>, string>
+        {
+          { x => Guid.TryParse(x.value?.ToString(), out Guid _) || String.IsNullOrEmpty(x.value?.ToString()), "Incorrect format of GenderId." },
```
Comment (MariyaSin) [inline src/UserService.Validation/User/EditUserRequestValidator.cs:207]: { x => string.IsNullOrEmpty(x.value?.ToString())? true :
          Guid.TryParse(x.value.ToString(), out Guid result)

File: src/UserService.Validation/User/EditUserRequestValidator.cs [hunk 18]
```diff
@@ -1,169 +1,244 @@
...
+      #endregion
+
+      #region Latitude
+
+      AddFailureForPropertyIf(
+        nameof(EditUserRequest.Latitude),
+        x => x == OperationType.Replace,
+        new Dictionary<Func<Operation<EditUserRequest>, bool>, string>
+        {
+          { x => double.TryParse(x.value?.ToString(), out double _) || String.IsNullOrEmpty(x.value.ToString()), "Incorrect format of Latitude." },
```
Comment (MariyaSin) [inline src/UserService.Validation/User/EditUserRequestValidator.cs:219]: same, use '? : ;'

File: src/UserService.Validation/User/EditUserRequestValidator.cs [hunk 19]
```diff
@@ -1,169 +1,244 @@
...
+      #endregion
+
+      #region Longitude
+
+      AddFailureForPropertyIf(
+       nameof(EditUserRequest.Longitude),
+       x => x == OperationType.Replace,
+       new Dictionary<Func<Operation<EditUserRequest>, bool>, string>
+       {
+          { x => double.TryParse(x.value?.ToString(), out double _) || String.IsNullOrEmpty(x.value.ToString()), "Incorrect format of Longitude." },
```
Comment (MariyaSin) [inline src/UserService.Validation/User/EditUserRequestValidator.cs:231]: same, use '? : ;'

File: src/UserService.Business/Commands/User/EditUserCommand.cs [hunk 20]
```diff
@@ -1,105 +1,134 @@
...
+
+      if ((!isAddEditRemoveUsers && userId != requestSenderId && !isAdmin) ||
+        (isActiveOperation is not null && !(isAddEditRemoveUsers || isAdmin)) || 
+        (isAdminOperation is not null && !isAdmin) || 
+        (isGenderOperation is not null && userId != requestSenderId))
+      {
+        return _responseCreator.CreateFailureResponse<bool>(HttpStatusCode.Forbidden);
+      }
+
+
```
Comment (MariyaSin) [inline src/UserService.Business/Commands/User/EditUserCommand.cs:82]: remove empty line

File: src/UserService.Business/Commands/User/EditUserCommand.cs [hunk 21]
```diff
@@ -1,105 +1,134 @@
...
+        if (!response.Body)
+        {
+          return _responseCreator.CreateFailureResponse<bool>(HttpStatusCode.NotFound);
+        }
+      }
+
+      await _cacheNotebook.RemoveAsync(userId);
+
+      response.Status = response.Errors.Any()
```
Comment (MariyaSin) [inline src/UserService.Business/Commands/User/EditUserCommand.cs:127]: remove it. You do not have any errors

**Review by MariyaSin — APPROVED** [review]

**Review by asyagraf — APPROVED** [review]

**Review by grinakate — APPROVED** [review]

**Pull request merged**